text/0000-alternate-registries.md

```diff
@@ -0,0 +1,120 @@
+- Feature Name: alternate-registries
+- Start Date: 2017-05-23
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Adding support for alternative crates.io servers to be used alongside the public crates.io server.
+
+# Motivation
+[motivation]: #motivation
+
+Currently cargo only has support for getting crates from a public server, this is fine for
+open source projects using Rust, however for closed source code this is problematic. Currently
+the only real option is to use Git repositories to specify the packages, but that means that
+all of the nice versioning and discoverability that cargo and crates.io provides is lost. I
+would like to change this so that it is possible to have a local crates.io server which private
+crates can be pushed to, plus still be able to make use of the public crates.io server as well.
+
+# Detailed design
+[design]: #detailed-design
+
+There are a number of different areas which will likely need to be tackled in order to fully
+support a local crates.io server in an enterprise. Below are some key areas:
+
+* Add support for alternate crates.io registries (this RFC)
+* Provide support for caching crates used on a crates.io proxy
+* Add support to crates.io to allow authentication with other OAuth providers than github
+* Support for private storage of crates on crates.io server rather than publicly available on S3
+
+Rather than trying to get everything agreed in a single RFC I would like for these to be
+dealt with as separate proposals and use this RFC as a stepping stone to be able to
+meaningfully start to use private registries.
+
+See https://github.com/rust-lang/cargo/pull/4036 for the code changes which this proposal was based on.
+
+## Cargo.toml config changes
+The following changes for Cargo.toml are proposed as part of this RFC:
+* Add an optional 'registry' field to be specified as part of the package dependencies
+* Add an optional 'registry' field to be specified in the package definition
```

Comment:
> supply a --source (or --registry) flag which tells cargo which registry to publish to

:+1:, in the given example the package could not specify a registry as the primary release channel is the default (or it could explicitly specify `crates.io` if it wanted), then setup the CI to do `cargo publish --registry https://internal.registry/` to override it if the tag is a prerelease version.